### PR TITLE
Adds owners tables and 	assigns owner in repo creation / update

### PR DESCRIPTION
**Pull request by srt32**
This PR updates the `RepoSynchronization` process and `Build`'s to assign the
`repo` to its parent `owner` (which is either a user or an organization). Tying
the `repo` to the `owner` is a preliminary step to having `owner` level
configuration.

New `repo`'s and updated `repo`'s (updated via the user clicking 'Refresh Repo
List' or via a build) will have their `owner`s populated.  We'll need to
backfill the `repo`'s that are not new and are not refreshed. I propose we merge
this PR as is and then have follow up work that adds an idempotent rake task
(which requires calls to GitHub) that will backfill `repo`'s missing
`owner_id`'s. Once the data is backfilled, we can apply a `null: false`
constraint to `repos.owner_id`.

Part of
https://trello.com/c/V0jY5LWs/452-migrate-existing-config-files-into-database

### CODE REVIEW

File: spec/factories.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
 FactoryGirl.define do
+  sequence(:github_id) { |n| n }
```
Comment (thorncp) [inline spec/factories.rb]: Is this block necessary?

**Comment by thorncp** [comment]
This looks good to me. Though, I regret not finishing up #577 yet. Conflicts abound :boom:

**Comment by croaky** [comment]
Does an owner mean an organization or a user? Should `repos.in_organization` get moved to `owners.organization`?

File: app/models/repo.rb [hunk 2]
```diff
@@ -1,8 +1,8 @@
 class Repo < ActiveRecord::Base
+  belongs_to :owner
```
Comment (croaky) [inline app/models/repo.rb]: > Order ActiveRecord associations alphabetically by attribute name.

https://github.com/thoughtbot/guides/tree/master/style/rails#rails

File: db/migrate/20150123224104_add_owner_to_repos.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class AddOwnerToRepos < ActiveRecord::Migration
+  def change
+    add_reference :repos, :owner, index: true
+    add_foreign_key :repos, :owners
```
Comment (croaky) [inline db/migrate/20150123224104_add_owner_to_repos.rb]: Interesting. Inspired by http://robots.thoughtbot.com/referential-integrity-with-foreign-keys ? I still haven't done this myself but I think I like it.

File: spec/factories.rb [hunk 4]
```diff
@@ -29,7 +32,7 @@
   end
 
   factory :user do
-    sequence(:github_username) { |n| "github#{n}" }
+    github_username { |n| "github#{n}" }
```
Comment (croaky) [inline spec/factories.rb]: The block looks like repetition of the `github_name` sequence.

```ruby
github_username { generate(:github_name) }
```

File: db/migrate/20141211224528_create_owners.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class CreateOwners < ActiveRecord::Migration
+  def change
+    create_table :owners do |t|
+      t.timestamps null: false
+      t.integer :github_id, null: false
+      t.string :github_name, null: false
```
Comment (croaky) [inline db/migrate/20141211224528_create_owners.rb]: Is it worth naming this column `github_username` to match the existing `users.github_username`?

File: app/models/owner.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class Owner < ActiveRecord::Base
+  has_many :repos, dependent: :destroy
+
+  def self.upsert(github_id:, github_name:)
```
Comment (croaky) [inline app/models/owner.rb]: How very Mongo of you.

File: app/models/owner.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    owner = find_or_initialize_by(github_id: github_id)
+    owner.github_name = github_name
+
+    if owner.changed?
```
Comment (croaky) [inline app/models/owner.rb]: Does a spec fail if this `if` statement is deleted? Is it needed?

File: app/services/repo_synchronization.rb [hunk 8]
```diff
@@ -19,12 +15,24 @@ def start
...
+  end
+
   def repo_attributes(attributes)
     {
       private: attributes[:private],
       github_id: attributes[:id],
       full_github_name: attributes[:full_name],
-      in_organization: attributes[:owner][:type] == ORGANIZATION_TYPE
+      in_organization: attributes[:owner][:type] == ORGANIZATION_TYPE,
+      owner: upsert_owner(attributes[:owner])
```
Comment (croaky) [inline app/services/repo_synchronization.rb]: > Use a trailing comma after each item in a multi-line list, including the last item.

https://github.com/thoughtbot/guides/tree/master/style/ruby
https://github.com/thoughtbot/guides/blob/master/style/ruby/sample.rb#L49

File: spec/models/repo_spec.rb [hunk 9]
```diff
@@ -1,6 +1,7 @@
 require "spec_helper"
 
 describe Repo do
+  it { should belong_to :owner }
```
Comment (croaky) [inline spec/models/repo_spec.rb]: After the associations are sorted by attribute name,

> Order validation, association, and method tests in the same order that they appear in the class.

https://github.com/thoughtbot/guides/tree/master/style/testing

File: spec/services/repo_synchronization_spec.rb [hunk 10]
```diff
@@ -1,91 +1,85 @@
...
       attributes = {
-        full_name: 'user/newrepo',
+        full_name: "user/newrepo",
         id: 456,
         private: true,
         owner: {
-          type: 'User'
+          type: "User",
+          id: 1,
+          login: "thoughtbot"
```
Comment (croaky) [inline spec/services/repo_synchronization_spec.rb]: `:sort`, trailing comma

Comment (croaky) [inline spec/services/repo_synchronization_spec.rb]: Same for other `owner` hashes later

File: spec/services/repo_synchronization_spec.rb [hunk 11]
```diff
@@ -103,20 +97,79 @@
...
+          attributes = {
+            full_name: "thoughtbot/newrepo",
+            id: repo_github_id,
+            private: true,
+            owner: {
+              type: "Organization",
+              id: owner.github_id,
+              login: owner.github_name
+            }
+          }
```
Comment (croaky) [inline spec/services/repo_synchronization_spec.rb]: Looks like a lot of repeated setup. I wonder if it'd be any easier to read if it were pulled out into a method. Don't know exactly what it'd look like, would need to extract then compare. Whatever you think.

File: app/models/owner.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+class Owner < ActiveRecord::Base
+  has_many :repos, dependent: :destroy
```
Comment (croaky) [inline app/models/owner.rb]: If deleting the owner deletes the repo, does it continue to cascade? Are the memberships, builds, and subscription deleted? Do they also need `dependent: :destroy` and foreign key constraints?

Comment (srt32) [inline app/models/owner.rb]: Good question! We don't currently destroy `owner`s anywhere via code. Assuming we do want to FK all the things, perhaps we should remove the `dependent: :destroy` here for now. If we do want to figure out the deletion use cases (cascading through all the associations) we can do that in another PR and apply all the FK's and `dependent: :destroys` consistently.

Thoughts?

Comment (croaky) [inline app/models/owner.rb]: My experience has been that if we don't consider the deletion case when we add the associations, we tend to forget to do it later and risk creating orphaned records. Even if we don't expose the UI, sometimes we'll accidentally do it through a rake task or console or something.

Doesn't matter to me whether that consideration comes in the form of a Trello card for later or handled in this pull request if we're confident that it should happen a certain way. My gut is that we probably want to remove the `dependent: :destroy` on `Owner has_many :repos` but definitely want to delete the subscription when a repo is deleted, so separate Trello card for that case.

File: app/models/owner.rb (hunk 7 above)
Comment (srt32) [inline app/models/owner.rb]: Nice, removed in cbb7dfa

File: app/models/owner.rb (hunk 12 above)
Comment (srt32) [inline app/models/owner.rb]: Ok, `dependent: :destroy` removed in e704e26 and a card for tracking here: https://trello.com/c/uTBir4jX/474-handle-deletion-cases

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (srt32) [inline db/migrate/20141211224528_create_owners.rb]: I think `username` could be confusing since an `owner` is either a `user` or an `organization`.

File: db/migrate/20150123224104_add_owner_to_repos.rb (hunk 3 above)
Comment (srt32) [inline db/migrate/20150123224104_add_owner_to_repos.rb]: Yup! And also inspired from `dependent: :destroys` not always working as well as giving us some more confidence when running SQL queries that delete / update data.

File: app/models/repo.rb (hunk 2 above)
Comment (srt32) [inline app/models/repo.rb]: Done in 8f54898

File: app/services/repo_synchronization.rb (hunk 8 above)
Comment (srt32) [inline app/services/repo_synchronization.rb]: Done in 8f54898

File: spec/factories.rb (hunk 1 above)
Comment (srt32) [inline spec/factories.rb]: Nice! Updated in a843dff

File: spec/factories.rb (hunk 4 above)
Comment (srt32) [inline spec/factories.rb]: Done in a843dff

File: spec/models/repo_spec.rb (hunk 9 above)
Comment (srt32) [inline spec/models/repo_spec.rb]: Done in e4a1fc6

File: spec/services/repo_synchronization_spec.rb (hunk 10 above)
Comment (srt32) [inline spec/services/repo_synchronization_spec.rb]: Done in e4a1fc6

File: spec/services/repo_synchronization_spec.rb (hunk 11 above)
Comment (srt32) [inline spec/services/repo_synchronization_spec.rb]: @croaky I took a stab in 04a1a0f.  Thoughts?

**Comment by salbertson** [comment]
@srt32, will all team members be able to update the configuration with this change and "owner level configuration"?

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (gylaz) [inline db/migrate/20141211224528_create_owners.rb]: How about `github_login`, since that would closer match what GitHub calls it?

File: spec/services/repo_synchronization_spec.rb [hunk 13]
```diff
@@ -94,29 +70,80 @@
...
+
+          synchronization.start
+
+          owner = Owner.find_by(github_id: owner.github_id)
+          expect(owner.repos.map(&:github_id)).to include(repo_github_id)
+        end
+      end
+    end
+
+    def stub_github_api_repos(repo_id:, owner_id:, owner_name:, private_repo: true, repo_name: "thoughtbot/newrepo")
```
Comment (gylaz) [inline spec/services/repo_synchronization_spec.rb]: How about?
```rb
def stub_github_api_repos(
  repo_id:,
  owner_id:,
  owner_name:,
  private_repo: true,
  repo_name: "thoughtbot/newrepo"
)
  attributes = ...
```

File: db/migrate/20141211224528_create_owners.rb [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+      t.index :github_name, unique: true
+      t.index :github_id, unique: true
+    end
```
Comment (gylaz) [inline db/migrate/20141211224528_create_owners.rb]: Should we also add a `organization` boolean field to this table?

File: app/models/owner.rb (hunk 7 above)
Comment (gylaz) [inline app/models/owner.rb]: I think the idea was to skip the trip to the db if the record didn't change, right?

**Comment by srt32** [comment]
@salbertson not exactly. This PR is step one for going down that road. This PR introduces owners (orgs or users) so that later we could associate style guides with those owenrs.

**Comment by salbertson** [comment]
@srt32, when would a style guide be associated with a user?

File: db/migrate/20141211224528_create_owners.rb (hunk 14 above)
Comment (srt32) [inline db/migrate/20141211224528_create_owners.rb]: I do think it makes sense to add that column here. Does that mean we would drop it from `repos`?

**Comment by srt32** [comment]
Good question. I was thinking that a `user` would have a style guide the same way an `organization` would have a style guide.

File: db/migrate/20141211224528_create_owners.rb (hunk 14 above)
Comment (gylaz) [inline db/migrate/20141211224528_create_owners.rb]: Yeah, that's what I was thinking.

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (srt32) [inline db/migrate/20141211224528_create_owners.rb]: Maybe just `login`?

File: spec/services/repo_synchronization_spec.rb (hunk 13 above)
Comment (srt32) [inline spec/services/repo_synchronization_spec.rb]: That approach is much nicer ;)

Comment (srt32) [inline spec/services/repo_synchronization_spec.rb]: Updated in 6882062

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (srt32) [inline db/migrate/20141211224528_create_owners.rb]: Done in dfa719f

**Comment by srt32** [comment]
I added the `owner.organization` column in dfa719f. My next commit will update that column.

File: db/schema.rb [hunk 15]
```diff
@@ -40,6 +39,17 @@
   add_index "memberships", ["repo_id"], name: "index_memberships_on_repo_id", using: :btree
   add_index "memberships", ["user_id"], name: "index_memberships_on_user_id", using: :btree
 
+  create_table "owners", force: :cascade do |t|
+    t.datetime "created_at",                   null: false
+    t.datetime "updated_at",                   null: false
+    t.integer  "github_id",                    null: false
+    t.string   "github_login",                 null: false
```
Comment (salbertson) [inline db/schema.rb]: Should we name this `github_username` to stay consistent with the users table?

File: app/jobs/buildable.rb [hunk 16]
```diff
@@ -3,6 +3,11 @@
 module Buildable
   def perform(payload_data)
     payload = Payload.new(payload_data)
+    Owner.upsert(
+      github_id: payload.repository_owner_id,
+      github_login: payload.repository_owner_name,
+      organization: payload.repository_owner_is_organization?
+    )
```
Comment (salbertson) [inline app/jobs/buildable.rb]: Why is the upsert done during a build and when repos are refreshed? Do we need to do it in both places?

File: app/models/payload.rb [hunk 17]
```diff
@@ -1,4 +1,5 @@
-require 'json'
+require "json"
+require_relative "../../lib/github_api"
```
Comment (salbertson) [inline app/models/payload.rb]: Why are these needed?

File: app/models/payload.rb [hunk 18]
```diff
@@ -35,10 +36,18 @@ def ping?
     data['zen']
   end
 
-  def repository_owner
+  def repository_owner_id
+    repository["owner"]["id"]
+  end
+
+  def repository_owner_name
```
Comment (salbertson) [inline app/models/payload.rb]: `repository_owner_username` to stay consistent?

File: db/schema.rb (hunk 15 above)
Comment (srt32) [inline db/schema.rb]: We went with `login` because we thought `username` might be confusing since an `owner` is either a `user` or an `org`.

https://github.com/thoughtbot/hound/pull/586#discussion_r23898835

File: app/models/pull_request.rb [hunk 19]
```diff
@@ -22,7 +22,7 @@ def comment_on_violation(violation)
   end
 
   def repository_owner
```
Comment (salbertson) [inline app/models/pull_request.rb]: `repository_owner_name` or `repository_owner_username` to make it clear a name is returned and not an owner object or something?

File: app/jobs/buildable.rb (hunk 16 above)
Comment (srt32) [inline app/jobs/buildable.rb]: I believe we do for 1) legacy reasons (we'll get backfilled data for free) and 2) we'll keep the active `owners` up to date this way (less stale data). Do you think the updates are unnecessary?

File: app/jobs/buildable.rb [hunk 20]
```diff
@@ -3,6 +3,11 @@
 module Buildable
   def perform(payload_data)
     payload = Payload.new(payload_data)
+    Owner.upsert(
+      github_id: payload.repository_owner_id,
+      github_login: payload.repository_owner_name,
+      organization: payload.repository_owner_is_organization?
```
Comment (gylaz) [inline app/jobs/buildable.rb]: `is_` usually comes first, in ruby `is_` is bypassed in favor of `?`. I'm not sure about `repository_owner_organization?`.

File: app/services/repo_synchronization.rb [hunk 21]
```diff
@@ -19,12 +13,25 @@ def start
...
+    @api ||= GithubApi.new(github_token)
+  end
+
   def repo_attributes(attributes)
     {
       private: attributes[:private],
       github_id: attributes[:id],
       full_github_name: attributes[:full_name],
-      in_organization: attributes[:owner][:type] == ORGANIZATION_TYPE
+      in_organization: attributes[:owner][:type] == GithubApi::ORGANIZATION_TYPE,
```
Comment (salbertson) [inline app/services/repo_synchronization.rb]: This comparison is done in a couple places. Can we get this from the payload object or return "GitHub repo" objects from `api.repos` that respond to `in_organization?`?

File: app/services/repo_synchronization.rb [hunk 22]
```diff
@@ -19,12 +13,25 @@ def start
...
+      owner: upsert_owner(attributes[:owner]),
     }
   end
+
+  def upsert_owner(owner_attributes)
+    Owner.upsert(
+      github_id: owner_attributes[:id],
+      github_login: owner_attributes[:login],
+      organization: owner_attributes[:type] == GithubApi::ORGANIZATION_TYPE
```
Comment (salbertson) [inline app/services/repo_synchronization.rb]: Yea, it would be nice if we could wrap these hash like data structures with an object with a nice interface.

File: app/models/payload.rb (hunk 17 above)
Comment (srt32) [inline app/models/payload.rb]: Running `rspec spec/models/payload_spec.rb` without the `require_relative` causes `  uninitialized constant GithubApi`. Is it worth keeping the requires so we can run individual testS?

File: app/services/repo_synchronization.rb (hunk 21 above)
Comment (srt32) [inline app/services/repo_synchronization.rb]: `in_organization` on the `repo` will go away once we have all the data backfilled on `owners` so I'd rather not invest in that change now. Thoughts?

File: app/services/repo_synchronization.rb (hunk 22 above)
Comment (srt32) [inline app/services/repo_synchronization.rb]: Do you mean something like `OwnerParameters` or something like it?

File: app/jobs/buildable.rb (hunk 20 above)
Comment (srt32) [inline app/jobs/buildable.rb]: I see what you mean. I enjoy the symmetry of these methods starting with `repositoy_owner_` but do find `repository_owner_organization?` funny to read. Maybe keep it as is?

File: app/models/payload.rb (hunk 18 above)
Comment (gylaz) [inline app/models/payload.rb]: Or maybe `repository_owner_login` to follow GitHub's naming?

File: spec/services/repo_synchronization_spec.rb [hunk 23]
```diff
@@ -94,29 +70,86 @@
...
+        synchronization = RepoSynchronization.new(second_user, "githubtoken")
 
         synchronization.start
 
         expect(second_user.reload.repos.size).to eq(1)
       end
     end
+
+    describe "repo owners" do
+      context "when the owner doesn't exit" do
```
Comment (salbertson) [inline spec/services/repo_synchronization_spec.rb]: "exist"?

File: app/services/repo_synchronization.rb [hunk 24]
```diff
@@ -19,12 +13,25 @@ def start
...
+  end
+
   def repo_attributes(attributes)
     {
       private: attributes[:private],
       github_id: attributes[:id],
       full_github_name: attributes[:full_name],
-      in_organization: attributes[:owner][:type] == ORGANIZATION_TYPE
+      in_organization: attributes[:owner][:type] == GithubApi::ORGANIZATION_TYPE,
+      owner: upsert_owner(attributes[:owner]),
```
Comment (gylaz) [inline app/services/repo_synchronization.rb]: Feels like this functionality is kind of hidden in the hash here. What do you think about calling it before the hash construction?

**Comment by salbertson** [comment]
@srt32, couple of comments/suggestions, looks good to me. :+1:

File: spec/factories.rb [hunk 25]
```diff
@@ -29,7 +33,7 @@
   end
 
   factory :user do
-    sequence(:github_username) { |n| "github#{n}" }
+    github_username { generate(:github_name) }
```
Comment (gylaz) [inline spec/factories.rb]: Should we just use `github_login` generator here and kill the `github_name` one?

File: spec/jobs/buildable_spec.rb [hunk 26]
```diff
@@ -1,7 +1,8 @@
-require 'fast_spec_helper'
-require 'app/jobs/buildable'
-require 'app/models/payload'
-require 'app/services/build_runner'
+require "spec_helper"
```
Comment (gylaz) [inline spec/jobs/buildable_spec.rb]: Why do we need `spec_helper` in this test?

File: spec/jobs/buildable_spec.rb [hunk 27]
```diff
@@ -1,7 +1,8 @@
...
+require "app/jobs/buildable"
+require "app/models/owner"
+require "app/models/payload"
+require "app/services/build_runner"
```
Comment (gylaz) [inline spec/jobs/buildable_spec.rb]: If we truly need `spec_helper` then these should be already loaded.

File: spec/factories.rb (hunk 25 above)
Comment (srt32) [inline spec/factories.rb]: hmmm, good question. `github_username` should be a subset of `github_login` so I think we should make this change.

File: lib/github_api.rb [hunk 28]
```diff
@@ -3,8 +3,10 @@
 require "active_support/core_ext/object/with_options"
 
 class GithubApi
-  SERVICES_TEAM_NAME = "Services"
+  ORGANIZATION_TYPE = "Organization"
   PREVIEW_MEDIA_TYPE = "application/vnd.github.moondragon-preview+json"
+  SERVICES_TEAM_NAME = "Services"
```
Comment (gylaz) [inline lib/github_api.rb]: Where is this used?

File: lib/github_api.rb [hunk 29]
```diff
@@ -3,8 +3,10 @@
...
+  USER_TYPE = "User"
```
Comment (gylaz) [inline lib/github_api.rb]: Do we need this? If it's only used in test, how about just hard-coding it there?

File: spec/models/payload_spec.rb [hunk 30]
```diff
@@ -73,7 +73,57 @@
...
+  end
+
+  describe "#repository_owner_is_organization?" do
+    context "when the repository owner is an user" do
+      it "returns false" do
+        payload_json = {
+          "repository" => {
+            "owner" => {
+              "id" => 1,
+              "type" => GithubApi::USER_TYPE
```
Comment (gylaz) [inline spec/models/payload_spec.rb]: Like, this can just be `"Not organization"`.

File: app/models/payload.rb (hunk 18 above)
Comment (srt32) [inline app/models/payload.rb]: Nice, done in 0c74b51

File: spec/services/repo_synchronization_spec.rb [hunk 31]
```diff
@@ -94,29 +70,86 @@
...
+            owner_id: owner_github_id,
+            owner_login: owner_login
+          )
+          synchronization = RepoSynchronization.new(user, "githubtoken")
+
+          synchronization.start
+
+          owner = Owner.find_by(github_id: owner_github_id)
+          expect(owner.github_login).to eq(owner_login)
+          expect(owner.repos.map(&:github_id)).to include(repo_github_id)
```
Comment (gylaz) [inline spec/services/repo_synchronization_spec.rb]: Is it safe to say `expect(owner.repos.map(&:github_id)).to eq([repo_github_id])`?

File: spec/services/repo_synchronization_spec.rb [hunk 32]
```diff
@@ -94,29 +70,86 @@
...
+            repo_id: repo_github_id,
+            owner_id: owner.github_id,
+            owner_login: owner.github_login
+          )
+          synchronization = RepoSynchronization.new(user, "githubtoken")
+
+          synchronization.start
+
+          owner = Owner.find_by(github_id: owner.github_id)
+          expect(owner.repos.map(&:github_id)).to include(repo_github_id)
```
Comment (gylaz) [inline spec/services/repo_synchronization_spec.rb]: Same here.

File: spec/services/repo_synchronization_spec.rb [hunk 33]
```diff
@@ -94,29 +70,86 @@
...
+        id: repo_id,
+        private: private_repo,
+        owner: {
+          id: owner_id,
+          login: owner_login,
+          type: "Organization",
+        }
+      }
+      resource = double(:resource, to_hash: attributes)
+      api = double(:github_api, repos: [resource])
```
Comment (gylaz) [inline spec/services/repo_synchronization_spec.rb]: `:github_api` -> `"GithubApi"`

File: app/models/payload.rb (hunk 17 above)
Comment (gylaz) [inline app/models/payload.rb]: The `require` should be in the test.

File: app/models/pull_request.rb (hunk 19 above)
Comment (srt32) [inline app/models/pull_request.rb]: Good idea, done in 94c22f6. This change touched a lot of files...

File: lib/github_api.rb (hunk 28 above)
Comment (srt32) [inline lib/github_api.rb]: I was just moving things around here. Looks like it's in `add_hound_to_repo` and 'github_api` (https://github.com/thoughtbot/hound/search?utf8=%E2%9C%93&q=SERVICES_TEAM_NAME). Perhaps keep it in `GithubApi`?

File: lib/github_api.rb (hunk 29 above)
Comment (srt32) [inline lib/github_api.rb]: Nice, done in c50f2fb

File: app/services/repo_synchronization.rb (hunk 24 above)
Comment (srt32) [inline app/services/repo_synchronization.rb]: Cool, done in c50f2fb

File: app/models/payload.rb (hunk 17 above)
Comment (srt32) [inline app/models/payload.rb]: Superb, done in c50f2fb

File: app/services/repo_synchronization.rb (hunk 22 above)
Comment (gylaz) [inline app/services/repo_synchronization.rb]: What if passed the `owner_attributes` into `upsert` and had it slice the right parameters out? Just a thought.

File: spec/jobs/buildable_spec.rb (hunk 27 above)
Comment (srt32) [inline spec/jobs/buildable_spec.rb]: `Owner` inherits from AR::Base so we need `spec_helper`. Removed the extra `requires` in 63d3952

File: spec/services/repo_synchronization_spec.rb (hunk 33 above)
Comment (srt32) [inline spec/services/repo_synchronization_spec.rb]: Done in 22db10d

File: spec/services/repo_synchronization_spec.rb (hunk 31 above)
Comment (srt32) [inline spec/services/repo_synchronization_spec.rb]: Nice, yup! 22db10d

File: app/services/repo_synchronization.rb (hunk 22 above)
Comment (srt32) [inline app/services/repo_synchronization.rb]: Thoughts on a potential fast follow refactor if we want to go that route? This PR is becoming large and long enough on its own ;)

File: db/schema.rb (hunk 15 above)
Comment (salbertson) [inline db/schema.rb]: `github_name`? Can be used in context of org or user.

File: app/services/repo_synchronization.rb (hunk 21 above)
Comment (salbertson) [inline app/services/repo_synchronization.rb]: Sure, makes sense.

File: app/services/repo_synchronization.rb (hunk 22 above)
Comment (salbertson) [inline app/services/repo_synchronization.rb]: Makes sense.

File: spec/factories.rb (hunk 25 above)
Comment (salbertson) [inline spec/factories.rb]: Haha, naming is hard.

File: db/schema.rb (hunk 15 above)
Comment (srt32) [inline db/schema.rb]: Agreed. Done in bb46f64 and e1828de

**Comment by srt32** [comment]
I squashed a bunch of intermediate commits to clean things up a bit but left the most recent commits there to make seeing the name changes easier. Another review would be appreciated cc/ @thorncp @croaky @gylaz @salbertson

File: db/schema.rb [hunk 34]
```diff
@@ -15,7 +15,6 @@
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
-  enable_extension "pg_stat_statements"
```
Comment (srt32) [inline db/schema.rb]: Hmmm, should this line have been removed?

File: spec/jobs/buildable_spec.rb (hunk 26 above)
Comment (salbertson) [inline spec/jobs/buildable_spec.rb]: @srt32, can we use `fast_spec_helper` here or is there something that's easier to test using `spec_helper`?

File: spec/models/payload_spec.rb [hunk 35]
```diff
@@ -73,7 +74,57 @@
...
+      }
+
+      payload = Payload.new(data)
+
+      expect(payload.repository_owner_id).to eq 1
+    end
+  end
+
+  describe "#repository_owner_is_organization?" do
+    context "when the repository owner is an user" do
```
Comment (salbertson) [inline spec/models/payload_spec.rb]: "a user"?

File: db/schema.rb (hunk 34 above)
Comment (salbertson) [inline db/schema.rb]: I'm not sure, probably not.

**Comment by salbertson** [comment]
@srt32, looks good to me! :+1:

File: db/schema.rb (hunk 34 above)
Comment (croaky) [inline db/schema.rb]: I think you can bring it back with:

```
rails dbconsole
CREATE EXTENSION pg_stat_statements;
\q
rake db:migrate
```

**Comment by teoljungberg** [comment]
Looks good to me aswell, this makes the refactoring to use the user's api token more easier. :+1:

**Comment by srt32** [comment]
Squashed and rebased in 67f6fa9

File: app/models/repo.rb [hunk 36]
```diff
@@ -1,9 +1,9 @@
 class Repo < ActiveRecord::Base
-  has_many :memberships
-  has_many :users, through: :memberships
   has_many :builds
-
+  has_many :memberships
+  belongs_to :owner
   has_one :subscription
+  has_many :users, through: :memberships
```
Comment (gylaz) [inline app/models/repo.rb]: `has_many`s Should be together?

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (gylaz) [inline db/migrate/20141211224528_create_owners.rb]: How about just `name`? I've actually aliased Repo's `full_github_name` to `name`, and always wonder, why we didn't just go with `name`.

File: app/models/repo.rb (hunk 36 above)
Comment (srt32) [inline app/models/repo.rb]: From https://github.com/thoughtbot/guides/tree/master/style/rails#rails

> Order ActiveRecord associations alphabetically by attribute name.

Am I misunderstanding?

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (srt32) [inline db/migrate/20141211224528_create_owners.rb]: I enjoy the symmetry with `github_id` but am fine with whatever.

File: spec/jobs/buildable_spec.rb (hunk 26 above)
Comment (gylaz) [inline spec/jobs/buildable_spec.rb]: I'm also curious why we need Rails for this spec.

File: spec/models/owner_spec.rb [hunk 37]
```diff
@@ -0,0 +1,39 @@
...
+        owner = create(:owner)
+        new_github_name = "ralphbot"
+
+        updated_owner = Owner.upsert(
+          github_id: owner.github_id,
+          github_name: new_github_name,
+          organization: true
+        )
+
+        expect(updated_owner.github_name).to eq new_github_name
```
Comment (gylaz) [inline spec/models/owner_spec.rb]: Do we need to call `updated_owner.reload` before the verification? What happens if `upsert` fails?

File: spec/models/payload_spec.rb [hunk 38]
```diff
@@ -1,6 +1,7 @@
-require 'fast_spec_helper'
+require "fast_spec_helper"
 require "attr_extras"
-require 'app/models/payload'
+require "app/models/payload"
+require "lib/github_api"
```
Comment (gylaz) [inline spec/models/payload_spec.rb]: This is a heavy requirement just for `GithubApi::ORGANIZATION_TYPE`. I'd favor just using a string.

File: app/models/repo.rb (hunk 36 above)
Comment (gylaz) [inline app/models/repo.rb]: When in doubt, change the style guide! :wink: https://github.com/thoughtbot/guides/pull/304

Let's keep it as it is for now.

Comment (srt32) [inline app/models/repo.rb]: you sly dog you.

File: spec/jobs/buildable_spec.rb (hunk 26 above)
Comment (srt32) [inline spec/jobs/buildable_spec.rb]: This spec is dependent on `Owner`, which inherits from `AR::Base`, so running the spec on its own breaks without Rails. We could create a `faksOwner` to get around this issue but it feels a bit heavy handed. Are there other stubbing techniques you'd recommend?

File: spec/models/owner_spec.rb (hunk 37 above)
Comment (srt32) [inline spec/models/owner_spec.rb]: Adding the `reload` has no impact. `Owner.upsert` calls `save` under the hood and then returns the `owner`, so whatever calls this code could handle errors however it likes. Does that make sense?

File: spec/models/payload_spec.rb (hunk 38 above)
Comment (srt32) [inline spec/models/payload_spec.rb]: Cool, updated in 2a293ec. Did you also mean to use the string within `Payload`?

File: spec/models/owner_spec.rb (hunk 37 above)
Comment (gylaz) [inline spec/models/owner_spec.rb]: So, `save` might fail silently. If `save` fails, should the `github_name` on the returned object still be set? I'm guessing so. Thus, `reload` would ensure it actually got updated in the db correctly. Maybe I'm overthinking this?

Comment (srt32) [inline spec/models/owner_spec.rb]: We changed `Owner.upsert` to use `save!` in 155085a. It makes sense to raise on failures because these cases should truly be exceptional (we are trying to insert a duplicate record where we should really be updating the record). 

@gylaz does this approach make sense to you?

File: db/migrate/20141211224528_create_owners.rb (hunk 5 above)
Comment (srt32) [inline db/migrate/20141211224528_create_owners.rb]: We went with `name` in f89eccd

File: app/models/style_guide/scss.rb [hunk 39]
```diff
@@ -53,7 +53,10 @@ def default_options
     end
 
     def default_config_file
-      DefaultConfigFile.new(DEFAULT_CONFIG_FILENAME, repository_owner).path
+      DefaultConfigFile.new(
+        DEFAULT_CONFIG_FILENAME,
+        repository_owner_name
+      ).path
```
Comment (salbertson) [inline app/models/style_guide/scss.rb]: Should this method be moved to `Base`?

File: spec/factories.rb [hunk 40]
```diff
@@ -53,4 +56,10 @@
     line_number 42
     messages ["Trailing whitespace detected."]
   end
+
+  factory :owner do
+    github_id
+    name { generate(:github_name) }
+    organization false
```
Comment (salbertson) [inline spec/factories.rb]: What do you think about removing this since the default is false?

File: spec/jobs/buildable_spec.rb (hunk 26 above)
Comment (salbertson) [inline spec/jobs/buildable_spec.rb]: Fine with me.

**Comment by salbertson** [comment]
@srt32, I really like the way this turned out, thanks for sticking with it. I have a couple questions, other than that is looks great. :+1:

File: app/models/payload.rb [hunk 41]
```diff
@@ -1,4 +1,5 @@
-require 'json'
+require "json"
+require "lib/github_api"
```
Comment (gylaz) [inline app/models/payload.rb]: Both of these get auto-loaded by Rails. For tests, we should move these to tests :wink:

File: app/models/repo.rb (hunk 36 above)
Comment (gylaz) [inline app/models/repo.rb]: I'm betting my option will win out!

File: app/models/style_guide/scss.rb (hunk 39 above)
Comment (gylaz) [inline app/models/style_guide/scss.rb]: :+1:

File: spec/models/style_guide/coffee_script_spec.rb [hunk 42]
```diff
@@ -93,7 +93,10 @@ class FooBar
         spy_on_file_read
         config_file = default_configuration_file(StyleGuide::CoffeeScript)
 
-        violations_in("var foo = 'bar'", repository_owner: "not_thoughtbot")
+        violations_in(
+          "var foo = 'bar'",
+          repository_owner_name: "not_thoughtbot"
```
Comment (gylaz) [inline spec/models/style_guide/coffee_script_spec.rb]: Shorten `not_thougthbot` to `foo`, and one line it?

Comment (gylaz) [inline spec/models/style_guide/coffee_script_spec.rb]: The spec description/context should be enough maybe.

**Comment by gylaz** [comment]
Blam! :+1: 

P. S. I really want to see the comment count of this PR to go to 200

File: spec/factories.rb (hunk 40 above)
Comment (srt32) [inline spec/factories.rb]: Great! Done in a619894

File: app/models/payload.rb (hunk 41 above)
Comment (srt32) [inline app/models/payload.rb]: Excellent. Done in a619894

File: spec/models/style_guide/coffee_script_spec.rb (hunk 42 above)
Comment (srt32) [inline spec/models/style_guide/coffee_script_spec.rb]: Sounds good. a619894

File: app/models/style_guide/scss.rb (hunk 39 above)
Comment (srt32) [inline app/models/style_guide/scss.rb]: @gylaz @salbertson thoughts on the implementation in 0d2cca3? Is there a better way to handle the subclasses' constants?

Comment (srt32) [inline app/models/style_guide/scss.rb]: In the name of getting this PR done, I'm going to :boot: on this one. I'll make a card to track this refactoring separately.
